### ISSUE TITLE
tf.name_scope with reuse

### ISSUE THREAD

**Issue by albertz** [tensorflow/tensorflow#6007]
I want to reuse a name scope which was created earlier with `tf.name_scope`. Normally, name_scope will create a new unique name in the current namespace. If it ends with "/", it will ignore the current namespace and use it as an absolute name scope but then it doesn't make it unique. I want a way that it just uses the name as I provide it in the current namespace.

Current code:
```
  @contextlib.contextmanager
  def name_scope(self, name):
      ...
      old_stack = self._name_stack
      if not name:  # Both for name=None and name="" we re-set to empty scope.
        new_stack = None
      elif name and name[-1] == "/":
        new_stack = name[:-1]
      else:
        new_stack = self.unique_name(name)
      self._name_stack = new_stack
      yield "" if new_stack is None else new_stack + "/"
```

I suggest something like:

```
  @contextlib.contextmanager
  def name_scope(self, name, reuse=False):
    if reuse:
      new_stack = self._name_stack + "/" + name
    else:
      ...
```

**Comment by mrry** [comment]
What would be the motivation for this feature? At the very least, its seems to conflict with the way `reuse` is meant in variable scopes, and that could be confusing for users.

**Comment by albertz** [comment]
I thought that `reuse` for variable scope has a very similar meaning ? (or even the same?)

The motivation is this:

```
with tf.name_scope("foo"):
   # sth

# ...

with tf.name_scope("foo", reuse=True):  # don't create "Foo_1" but use "Foo" name scope
  # sth
```

**Comment by mrry** [comment]
Wouldn't the following work just as well?

```python

with tf.name_scope("foo") as scope:
   # sth

# ...

with tf.name_scope(scope):
  # sth
```

**Comment by albertz** [comment]
@mrry Yes but than I need keep track about all the scopes which is not what I want. In my specific case, in one place of the code, I create all the variables with a deeper scope hierarchy and in another separated place, I define my computation graph which uses these vars, and I want it to be in the same scopes.

**Comment by mrry** [comment]
I have no idea if that's something that we'd like to support. I'll defer to @lukaszkaiser, who added the equivalent feature in `tf.variable_scope()` and has been simplifying the existing scopes over the last few months.

**Comment by lukaszkaiser** [comment]
We were considering this long time ago as the first version of variable_scope. Then the reuse checks came and so on -- it basically evlolved into variable_scope. Currently, we recommend everyone to use variable_scope and not use name_scope except for internal code and libraries. Wouldn't just using variable_scope solve your problem?

**Comment by albertz** [comment]
I just checked the code of variable_scope, and it seems as if even with reuse=True, it would not reuse the name_scope. The reuse option doesn't seem to have an effect on the name_scope. It's basically this code, if I use `tf.variable_scope("foo", reuse=True)`:

```
  with g.as_default():
      name_scope = name_or_scope
      with ops.name_scope(name_scope) as cur_name_scope:
          with _pure_variable_scope(
              name_or_scope,
              reuse=reuse,
          ...
```
So, as it is right now, even variable_scope would not reuse the name scope, but I want to reuse the name scope. Is that even the intended behavior of variable_scope or a bug?

**Comment by lukaszkaiser** [comment]
The idea of reusing a name_scope is a bit strange, because every op in the graph has to have a unique name anyway in the end. So you're only pushing back where the suffix will be added, it will go at the end od the op-name instead of the scope name. You're aware of that, right? I just want to make sure that that's really the goal you want to achieve.

If that's the goal, then you can do this with variable_scope and its .original_name_scope property.

```
# First time just open variable_scope.
with tf.variable_scope("test") as scope:
  # do things, name_scope = "test"
  ...
# Next time open name_scope with scope.original_name_scope.
with tf.name_scope(scope.original_name_scope):
   # we're back in the name_scope as before
```

Does that help?

**Comment by albertz** [comment]
Yes, that's my goal. Your suggestion is basically what @mrry suggested but this solution is not optimal because I would need to keep the `scope` variable around for all scopes where I want to do this, which is for almost all scopes. This is because I create all the variables for my model at one place, with a specific name scope hierarchy, and then at some later time, I create the computation graph, and I want it to use the same name scope hierarchy.

**Comment by lukaszkaiser** [comment]
I see, but your use case is very specific and very strange. I think what you want deep down might be an object-oriented layer API like in Keras, and it is being done in tf.layers and will be part of contrib. In its technical aspect it relies on storing the scope as part of an object, as discussed above. In all cases, I think we can close this issue.

**Comment by npfp** [comment]
Hi,

As a matter of fact, it appears that I had the same use case than albertz 
- variables created and stored in one place,
- various operations defined later and if possible in the same scope than the variables.

I ended up independently to the same conclusion that the scope object had to be stored in the same place than the variables.

It looks like it's not the right way to work with tensorflow variables and so I wonder if you could point me to some guidelines/good practices to structure models. [I'm currently following the practices described here: https://danijar.com/structuring-your-tensorflow-models/, but it doesn't work so well as mentioned above when you want to use the same scope for your operations).

Thx!

**Comment by lukaszkaiser** [comment]
I think there are 2 good ways to structure TF models.

* The functional / scoping way. Everything is a function, create variables inside (don't worry about creating them beforehand). I use this a lot for research as it makes it super-easy to just put a new layer here and there. It is the basis of [tensorflow-slim](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/slim) and there are many examples in the [slim models library](https://github.com/tensorflow/models/tree/master/slim), e.g., take a look at [AlexNet](https://github.com/tensorflow/models/blob/master/slim/nets/alexnet.py).

* The OO way. Every layer is an object, you need to define composition models. We're still working on putting it in core TF (see tf.layers implementation, a lot is there already). But, in essence, it's the same ideas as [Keras](https://github.com/fchollet/keras) and its conventions.

Hope that helps!

**Comment by npfp** [comment]
Thx, it definitely  helps. I appreciate!

**Comment by Faur** [comment]
I have another use-case for the having multiple variables have the same name. (I might be completely misunderstanding something, any help is appreciated)

I want to track some variables in TensorBoard. I am saving the validation loss and training loss (and other things) in separate directories, such that TensorBoard thinks they are from separate runs. The only problem with this approach is that duplicate names get a '`_1`' appended to the name, so they don't show up in the same plot.

**Comment by ppwwyyxx** [comment]
@Faur Your issue is more related to #6150 than this one.

**Comment by ngc92** [comment]
What about the following use case:
I am observing the gradients in my model with summary operations, but there are two separate parts of the codes where gradients are calculated. However, I would like the summaries to go into the same name scope.

**Comment by albertz** [comment]
Note that I have implemented this function myself [here in `reuse_name_scope`](https://github.com/rwth-i6/returnn/blob/master/TFUtil.py#L491). This is a bit ugly right now because I wanted it to support both the variable scope and the name scope.

**Comment by javidcf** [comment]
I found out that at least in the current version (1.2), if you call `tf.name_scope` with a string ending in `/` it will _not_ add a suffix even if it already exists. Although it is not documented, I hope the developers let it be like that; for me, it seems fairly clear that saying `tf.name_scope("my_scope/")` means that I explicitly want to use the scope `my_scope` exactly.

**Comment by albertz** [comment]
A string ending with a slash will be interpret as an absolute name also,
not relative.

This does not work however for variable_scope.


Am 12.07.2017 19:13 schrieb "Javier Dehesa" <notifications@github.com>:

**Comment by javidcf** [comment]
@albertz Thanks for the info. I guess the complete snippet to reuse an existing name scope would then be:

```
current_name_scope = tf.get_default_graph().get_name_scope()
if len(current_name_scope) > 0:
    current_name_scope += "/"
with tf.name_scope("{}my_scope/".format(current_name_scope)):
    # ...
```

Although as you said `tf.variable_scope` / `tf.get_variable` work differently. In fact, if you do:

```
with tf.name_scope("a"):  # or tf.variable_scope("a")
    pass

with tf.variable_scope("a"):
    v = tf.get_variable("v", 0)
    v2 = v * 2

print(v.op.name, v2.op.name)
```

You get `a/v a_1/mul`, which I think is rather confusing (I'd expect that everything I create within a scope actually ends up in the same scope).

**Comment by albertz** [comment]
@javidcf In case you are interested, I implemented `reuse_name_scope` [here](https://github.com/rwth-i6/returnn/blob/master/TFUtil.py), which tries to handle both the name scope and the variable scope. It's a bit confusing anyway, as the `tf.variable_scope` will implicitly also set the name scope but not the other way around. As @lukaszkaiser maybe implied, name scope might go away at some point, or they will be both merged together. I don't really see why there is the distinction at all.

**Comment by ngc92** [comment]
I think the distinction makes sense as soon as you replicate a part of a network, but feed it different inputs. For example, if you want to re-input adversarial examples back into your network to train it to classify those correctly too, you want to reuse the variable scope of the original model, but put all your operations into a separate scope (at least if you want to inspect your graph at some point and want it to be structured)

**Comment by cheind** [comment]
@albertz, @javidcf I was following your discussion here, as I've been also interested in re-opening name scopes. Here's a bit of warning though (without having studied your implementations). When you add `/` to your name scope names, TF will interpret the name scope as absolute path, ignoring any parental name-scopes that you are currently in. This can be quite suprising. Here's a short example that illustrates the issue.

```python
import tensorflow as tf

with tf.Graph().as_default() as g:
   with g.name_scope('parent'):
       with g.name_scope('nested'):
           x = tf.placeholder(dtype=tf.float32, name='x')
           print(x.name)
       with g.name_scope('nested'):
           y = tf.placeholder(dtype=tf.float32, name='y')
           print(y.name)
       with g.name_scope('nested/'):
           y = tf.placeholder(dtype=tf.float32, name='y')
           print(y.name)
   
# Gives
# parent/nested/x:0
# parent/nested_1/y:0
# nested/y:0
```
Although no suffix was added to `nested` name scope, it lost the parental relationship to `parent`.
